### ISSUE TITLE
F401 rule has a bug

### ISSUE THREAD

**Issue by seriaati** [astral-sh/ruff#13181]
I'm not exactly sure if this is a bug.
Consider the following file structure:
```
my_view_items/
├─ my_button.py
components.py
my_view.py
```
With the following code:
```py
# components.py

from typing import Generic, TypeVar

V_co = TypeVar("V_co", bound="View")


class View:
    pass


class Button(Generic[V_co]):
    pass
```
```py
# my_view.py

from .components import View


class MyView(View):
    pass
```
```py
# my_view_items/button.py

from typing import TYPE_CHECKING

from ..components import Button

if TYPE_CHECKING:
    from ..my_view import MyView # F401 is raised here.


class MyButton(Button["MyView"]):
    pass
```

The example can probably be more simplified to reproduce this issue, but this is something that I encountered in one of my projects, so I decided to just simplify the code (from my project).
Basically, typing using quotes for generics would cause F401 (unused import) to be triggered. But the import is actually needed.
In addition, the presence of `from __future__ import annotations` doesn't affect whether F401 is raised.

**Comment by charliermarsh** [comment]
This is similar to https://github.com/astral-sh/ruff/issues/9298. We can't detect arbitrary user-defined generics right now. You can tell Ruff that a class should be considered generic via https://docs.astral.sh/ruff/settings/#lint_pyflakes_extend-generics for now, and we plan to improve it in the future.
